keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,579 @@
+We propose to introduce `.spec.jitter` which is a percentage of the time delta to the next schedule. We propose to cap it to 50%. There is also a [community request](https://github.com/kubernetes/community/issues/2440) for this.
+
+```golang
+delta = nextScheduleTime - currentTime
+jitter = delta*cronjob.Spec.Jitter/100
+nextScheduleTime += jitter
+```
+
+### Fix applicable open issues
+These are the [current](https://github.com/kubernetes/kubernetes/issues/82659) list of issues that are being targeted for GA. 
+
+- [Updating a cronjob causes jobs to be scheduled retroactively](https://github.com/kubernetes/kubernetes/issues/63371)
+- [CLI: Updated CronJob Schedule Missing from Dry Run](https://github.com/kubernetes/kubernetes/issues/73613)
+- [Kubernetes CronJob pods is not getting clean-up when Job is completed](https://github.com/kubernetes/kubernetes/issues/74741)
+- [Infinite ImagePullBackOff CronJob results in resource leak](https://github.com/kubernetes/kubernetes/issues/76570)
+- [Cronjob `spec.schedule` cannot be change when `spec.schedule` value not `"` or `’`](https://github.com/kubernetes/kubernetes/issues/78646)
+- [Kubelet CPU/Memory Usage linearly increases using CronJob](https://github.com/kubernetes/kubernetes/issues/64137)
+- [Stopping cluster overnight prevents scheduled jobs from running after cluster startup](https://github.com/kubernetes/kubernetes/issues/42649)
+
+### Scale Targets for GA
+
+The scale targets for GA of CronJob are defined by the same [API call latency
+SLIs/SLOs as the Kuberetes native types](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md#api-call-latency-slisslos-details).
+
+The targets are defined by the below suggested maximum limits, which are organized the same way as the [Kubernetes native type thresholds](https://github.com/kubernetes/community/blob/master/sig-scalability/configs-and-limits/thresholds.md#kubernetes-thresholds).
+
+#### CronJob Limits
+There should be nothing in the implementation that limits CronJobs per namespace. Overall clusterwide limits of CronJob are important. Cluster wide limits for CronJob should be storage bound since it shares the storage space with all other objects. Determining the appropriate storage limit for a cluster is out-of-scope for this document. We would recommend having CronJob use not more than 15% of the etcd storage. This allows space for Jobs created by CronJobs with the default successfulJobsHistoryLimit of 3 and for other objects.
+
+#### Frequency of launched jobs
+The number of CronJobs is also sensitive to the API server QPS and the schedule of the individual CronJobs. This translated to the frequency of launched jobs. We could have large number of CronJobs with a spread of schedule that doenst stress the Job API. At the same time we could have a small number of CronJobs that schedule synchronously stressing the Jobs API. The design must be able to easily saturate the API server QPS.
```

Comment:
Should we add something like "we will rate-limit job creation to N QPS"?

As of now we have it out-of-the-box by QPS limits, with priority and fairness it will no longer be true, but for cron-jobs we may actually want to leave this, to avoid infinite accumulation of jobs...

Reply:
BTW - this infinite accumulation of jobs coming from cron-jobs seem like a failure mode i wasn't really realizing of and has a potential to fully break your cluster (we will be creating jobs faster than we are able to run and delete them, leaning to constant grow of etcd memory usage, finally firing the "no-space" alarm and making etcd in read-only mode, causing an outage of your cluster).

That makes me thing, that maybe we should also add a metric for sth like "number of existing jobs owned by cron-jobs".

Reply:
We can rely on https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#eventratelimit for Jobs as set by the administrator instead of internally ratelimiting in the CronJob controller.

We should not have infinite accumulation since we have SuccessfulJobsHistoryLimit and FailedJobHistoryLimit.

Reply:
Iam mentioning explicitly that the user can set the ratelimits